File: shell/preload/scoped_env.py
import json
import os

directory = os.environ.get('QL_TASK_ENV_SNAPSHOT')
snapshot = None
if directory:
    with open(os.path.join(directory, 'snapshot.json'), encoding='utf-8') as source:
        snapshot = json.load(source)


def apply_scoped_environment():
    if snapshot is None:
        return
    for name in snapshot['unset']:
        os.environ.pop(name, None)
    os.environ.update(snapshot['overlay'])

File: shell/preload/scoped-env.js
const fs = require('fs');
const path = require('path');
const directory = process.env.QL_TASK_ENV_SNAPSHOT;
const snapshot = directory ? JSON.parse(fs.readFileSync(path.join(directory, 'snapshot.json'), 'utf8')) : null;
exports.directory = directory;
exports.apply = function applyScopedEnvironment() {
  if (!snapshot) return;
  for (const name of snapshot.unset) delete process.env[name];
  for (const [name, value] of Object.entries(snapshot.variables)) process.env[name] = value;
};
